### PR TITLE
thrift: add support for case-sensitive header keys

### PR DESCRIPTION
**Pull request by rgs1**
Based off #20596, this preserves the current behavior by default.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>
Co-authored-by: fishy

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @fishcakez @JuniorHsu

**Comment by rgs1** [comment]
cc: @davinci26 @fishy

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by wbpcode** [comment]
Altough, by this way we can insert some case insensitive header key to the Http::HeaderMap, here are some others problem.

The headers will be used in the header matching or getting cluster with cluster_header or some other scenes. 
And the key in these config will always be convert to lower case. **Because the `get` API of `Http::HeaderMap` only accept key of type `Http::LowerCaseString`.** This will result in some weird results.
For example,  if we set `cluster_header` to "Key", and there is related "Key" and "Value" in the headers. But we can never get the valid header value.

I think the key point is that we shouldn't use the `Http::HeaderMap` as the header map type of thrift. The `Http::HeaderMap` is designed for HTTP (lower case key, multiple value for same key), and headers of thrift should be a normal string map (ref https://github.com/apache/thrift/blob/master/lib/cpp/src/thrift/transport/THeaderTransport.h).

**Comment by wbpcode** [comment]
> I think we can document all of these limitations and still move forward with this. Speaking of HTTP, we do support case preservation there (see https://github.com/envoyproxy/envoy/pull/15619).

Yeah, but the seaching of header key will no be influenced by this preservation. And the approach in this PR does.

IMO, I think it would be better to replace `Http::HeaderMap` in the thrift proxy with `absl::flat_hash_map<std::strinig, std::string>`.
Or we can use the approach in the  https://github.com/envoyproxy/envoy/pull/15619 to preserve the case-sensitive key.

**Comment by rgs1** [comment]
> > I think we can document all of these limitations and still move forward with this. Speaking of HTTP, we do support case preservation there (see #15619).
> 
> Yeah, but the seaching of header key will no be influenced by this preservation (#15619). And the approach in this PR does.
> 
> IMO, I think it would be better to replace `Http::HeaderMap` in the thrift proxy with `absl::flat_hash_map<std::strinig, std::string>`. Or we can use the approach in the #15619 to preserve the case-sensitive key.

Replacing `Http::HeaderMap` sounds tempting, but rely on it for some things:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/thrift_proxy/router/router_impl.cc#L117

So it sounds like the proper way of doing this would be he approach in the #15619 (at the cost of using more mem).

**Comment by wbpcode** [comment]
Yeah, the approach in the #15619 is best way for now. 

(May be we can try to make the http header match utility a template, then it can be easily reused by other type headers (e.g. `absl::flat_hash_map<std::strinig, std::string>`. I will have a try. Because the dubbo also need this.)

**Comment by rgs1** [comment]
@wbpcode ok I pushed something similar to #15619 but less complex (e.g.: the formatter is hard coded because, at least right now, we don't need a whole new plugging system for headers formatting in Thrift). Let me know what you think, thanks!

**Comment by wbpcode** [comment]
@rgs1 Great. I think it's good enough for now. The case sensitive keys can be preserved and the headers matching will not be affected.

**Comment by rgs1** [comment]
Unrelated error:

```
emcc: error: '/b/f/w/external/emscripten_bin_linux/bin/clang -target wasm32-unknown-emscripten -DEMSCRIPTEN -D__EMSCRIPTEN_major__=3 -D__EMSCRIPTEN_minor__=1 -D__EMSCRIPTEN_tiny__=7 -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=/b/f/w/external/emscripten_bin_linux/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -DNO_FSEEKO -I/b/f/w/bazel-out/wasm-fastbuild-ST-c13b05b68ca8/bin/external/envoy/bazel/foreign_cc/zlib.build_tmpdir -I/b/f/w/external/net_zlib --sysroot=/b/f/w/external/emscripten_bin_linux/emscripten/cache/sysroot -fdiagnostics-color -fno-strict-aliasing -funsigned-char -no-canonical-prefixes -fomit-frame-pointer -O0 -Wall -iwithsysroot/include/c++/v1 -iwithsysroot/include/compat -iwithsysroot/include -isystem /b/f/w/external/emscripten_bin_linux/lib/clang/15.0.0/include -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -Werror -O3 -fexceptions -MD -MT CMakeFiles/zlib.dir/gzread.o -MF CMakeFiles/zlib.dir/gzread.o.d -c /b/f/w/external/net_zlib/gzread.c -o CMakeFiles/zlib.dir/gzread.o' failed (returned 1)
```

File: source/extensions/filters/network/thrift_proxy/metadata.h
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/metadata.h:27]: @mattklein123 should we move this copy/pasted class to some common path?

**Comment by rgs1** [comment]
@zuercher I think now that we have an agreement with @wbpcode this is ready for review. Thanks!

File: source/extensions/filters/network/thrift_proxy/metadata.h
Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/metadata.h:27]: Ok cool, added a TODO for now so we can avoid and unblock @fishy.

**Comment by rgs1** [comment]
@fishy any chance you can give this a try before we merge?

**Comment by marcoferrer** [comment]
I can give it a shot. One question I have is how can I reference an image with a build of this PR?

**Comment by rgs1** [comment]
> I can give it a shot. One question I have is how can I reference an image with a build of this PR?

I am afraid we might not have that...

**Comment by rgs1** [comment]
> > I can give it a shot. One question I have is how can I reference an image with a build of this PR?
> 
> I am afraid we might not have that...

As in, afaik there are no images published for open PRs. But if @zuercher is happy with the state of this, I think there might be an image created after it lands in main.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc
Comment (fishy) [inline source/extensions/filters/network/thrift_proxy/header_transport_impl.cc:158]: yes, the removal is also a violation of the spec, see this commit on the clarification of the spec: https://github.com/apache/thrift/commit/e165fa3c85d00cb984f4d9635ed60909a1266ce1

this should be moved to be before the `StrReplaceAll` call.

Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc:158]: cc: @wbpcode

Comment (wbpcode) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc:158]: I agree that we should move the formatter code before StrReplaceAll call.

File: docs/root/version_history/current.rst
Comment (zuercher) [inline docs/root/version_history/current.rst:45]: I think this has to move to the YAML file before we can merge.

File: source/extensions/filters/network/thrift_proxy/header_transport_impl.cc
Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc:158]: I guess we're waiting on this, but otherwise seems good to me.

Comment (fishy) [reply source/extensions/filters/network/thrift_proxy/header_transport_impl.cc:158]: to clarify I'm totally ok if we merge the current version first, cut a release, and address this issue in a follow up PR later.

although they are both spec violations, case sensitivity is far more severe/common than those special characters. fixing wrong case sensitivity will fix probably 99% of the spec violations. so if fixing this extra case is too complicated it might worth it to get the other fix out first. (but I'll leave that to y'all to decide)

**Comment by rgs1** [comment]
@htuch for API.

**Comment by rgs1** [comment]
@wbpcode for another pass.

File: api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto
Comment (zuercher) [inline api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:122]: Is this comment correct? 

Internally (e.g. for route machine, substation formatters, etc) Envoy will still use case-insensitive header keys stripped of NUL, CR and LF. My understanding is that enabling this flag will cause forwarded header keys to retain their original formatting.

**Review by zuercher — COMMENTED** [review]
Code looks good. I'd like to see the comment on the API field clarified (or else I'm deeply confused 😃 ).

File: api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto
Comment (rgs1) [reply api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:122]: Vaguely correct, if anything. How about we go with something like the following (reusing most of your words):

If set to true Envoy will forward headers retaining their original formatting while internally route matching and other filters will treat them as case-insensitive and they'll also be stripped of NUL, CR and LF.

cc: @wbpcode @fishy @fishcakez

Comment (fishcakez) [reply api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:122]: should we rename this to `header_keys_preserve_case` and then comment like `If set to true, Envoy will preserve the case of Thrift header keys instead of serializing them to lower case as per the default behavior`.

The header keys aren't actually case sensitive internally because we're using the http matchers etc?

**Comment by wbpcode** [comment]
The code is LGTM. Thanks for your great work.

cc @htuch Can you take a look for the API when you have free time? Thanks.

File: api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto
Comment (htuch) [reply api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:122]: Agree with both suggestions.

Comment (rgs1) [reply api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto]: Cool, thanks all. Updated.

**Review by fishcakez — APPROVED** [review]

File: api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:123]: Nit: `as mandated..`

**Review by htuch — COMMENTED** [review]
/lgtm api

**Review by wbpcode — DISMISSED** [review]
LGTM. Thanks for your great work. cc @zuercher for final pass or merge.

**Comment by wbpcode** [comment]
cc @htuch could your give another lgtm for the API? Thanks.

**Comment by rgs1** [comment]
@zuercher this is ready for another pass (the macos failure was something else, possibly transient).

**Review by zuercher — APPROVED** [review]
Looks good!

**Pull request merged**
